### ISSUE TITLE
Problem with the import of a "Test"

### ISSUE THREAD

**Issue by yfrindel** [solidify/jira-azuredevops-migrator#62]
I have a problem with the type "Test". When I try to import a "Test" from Jira, the workitem is empty if it's mapped with the type "Test Case" of VSTS. Could you help me please ?

**Comment by MOlausson** [comment]
Can you share your mapping, I'm sure it's correct but want to make sure we start looking from the right angle.

**Comment by yfrindel** [comment]
yes sure, thx
[config-scrum.zip](https://github.com/solidify/jira-azuredevops-migrator/files/3310851/config-scrum.zip)

**Comment by yfrindel** [comment]
I need especially some things like the "history" or the "test step"

**Comment by yfrindel** [comment]
I can extract now the "test step" in a XML file, like that : 
![image](https://user-images.githubusercontent.com/22955005/60091993-2376bd80-9746-11e9-8ca3-9f36f6d89919.png)

**Comment by machinatus** [comment]
I have also run into this issue for a number of the fields from a Jira Test record. I am attaching my config file as well.

For the Summary/Title, I am not sure what the standard {"mapper": "MapTitle"} includes, so I also tried adding a separate {"source": "summary", "target": "Microsoft.VSTS.TCM.AutomatedTestName", "for": "Test Case"} section but to no avail.

Note that I am migrating to an Agile template, not Scrum, but the issue appears to be with the export not the import anyway.

[config_tests.zip](https://github.com/solidify/jira-azuredevops-migrator/files/3331550/config_tests.zip)

Thanks for any possible assistance.

**Comment by machinatus** [comment]
If this cannot be fixed for Test issues from Jira, what would the work around be? To export using code (as shown by yfrindel) and to import to ADO using Excel (https://docs.microsoft.com/en-us/azure/devops/boards/backlogs/office/bulk-add-modify-work-items-excel?view=azure-devops)?

**Comment by MOlausson** [comment]
@machinatus we've had limited bandwidth last month but I'll take a look at this now. It sounds like a gap that needs to be fixed. Hope to have an update for you tomorrow.

**Comment by machinatus** [comment]
Thank you for responding. Since there has been no further update, would it be correct to assume that I should look for alternate methods than the Solidify tool? I am under some time pressure to resolve this, so while I like the tool and want to support it, I cannot allow this issue to stop my migration project. :)

**Comment by MOlausson** [comment]
Hi @machinatus, sorry but we started with some other issue so did not get to look at this. But will now. Can you help me repro, how are you creating the test cases (tests) in Jira? Using a custom issue type or an extension? Great if you can share some details regarding the issue definition.

**Comment by machinatus** [comment]
Thanks. I am not an active Jira participant in the development project. The team says that they are using the Zephyr extension, which I guess must modify the Test record object.

**Comment by MOlausson** [comment]
Got it, I just installed that one so will give it a try. Should be enough for me to repro and the you can help me verify. I'll post an update later today.

**Comment by MOlausson** [comment]
So I looked at the data we get back from the Jira REST api and unfortunately we don't get the specifics from the Zephyr test case. What I can see (from a very quick glance) is that the extension makes an external call to get the custom data (like test steps) and that's not something we currently support. It makes for an interesting extension so will add that to the backlog but not likely something we will do in the next couple of weeks.

Perhaps you can export the test cases to excel and then use https://docs.microsoft.com/en-us/azure/devops/test/reference-qa?view=azure-devops&viewFallbackFrom=vsts#q-is-there-a-way-to-quickly-add-multiple-test-cases-at-the-same-time to import it to Azure DevOps as a short-term solution. The other issue types (excluding extension) can be exported/imported using our tool of course.

**Comment by MOlausson** [comment]
@yfrindel are/were you also using the Zephyr test extension?

**Comment by yfrindel** [comment]
Indeed, we use the zephyr extension. And with this XML export, I'm able now to import/export all the fields that I want (spec steps for the tests cases, Integrated build, found build, etc...). I have just still some issues with some "histories" not correctly imported, I don't know why...

**Comment by MOlausson** [comment]
Ok, got it. I've been in contact with Zephyr support and they have an api we can use to get the test case information, now we have to look at a design for this type of extension. So short-term we're not supporting 3rd party extensions but I hope we will soon. If you feel the history issues are related to our tool then feel free to open a separate issue for that.

**Comment by brbones** [comment]
I am running into this same issue and exporting from Jira to excel does not export the needed data either...so i am unable to import all that i need to VSTS.

**Comment by machinatus** [comment]
@brbones: For what it's worth, getting the Test Cases which have Zephyr additional info out of Jira required assistance from the Zephyr support team initially. I took the resulting CSV file and massaged it into what I wanted to paste into the Grid View in Azure DevOps (https://docs.microsoft.com/en-us/azure/devops/test/reference-qa?view=azure-devops#q-is-there-a-way-to-quickly-add-multiple-test-cases-at-the-same-time).

**Comment by MOlausson** [comment]
@brbones I understand your consern here, we're investigating how we can add support for 3rd party extensions. No update there yet so short term I would suggest you look into the approach @machinatus has used.

**Comment by kmuthukumar-rp** [comment]
Coping some info that I pasted on one of the other ticket regarding Zephyr, as it may be helpful. 

@MOlausson @yfrindel @machinatus @brbones If it helps any, you would need to hit Zephyr API to get the test steps values. Currently the way I had dealt with it is:  In JiraItem.ExtractFields(), for testcase issuetype and customfield teststep, download and transform the test steps values into an array of objects by hitting the Zephyr API `rest/zapi/latest/teststep/{issueId}.` We might be able to write a mapper to map teststeps, but I couldnt get it to working, so did it this way.

Then save the test step values for each testcase at the end of ImportCommandLine.ExecuteMigration by maintaining an array of wi.Id's when test item gets created (only once, not for every revision) and then process them at the end. 

> First get the saved TestCase by,

```
var testService = Collection.GetService<ITestManagementService>();
            var testProject = testService.GetTeamProject(_project);
            ITestCase testCase = testProject.TestCases.Find(wiId);
            return testCase;
```

> Then AttachSteps and save:

```
ITestStep step = testCase.CreateTestStep();
step.Title = stepTitle;
 step.Description = stepTitle;
step.ExpectedResult = stepElement.Element(Constants.TESTSTEP_RESULT).Value;
testCase.Actions.Add(step);
```
I am still missing to get teststep attachments which is avl with the API call: `/rest/zapi/latest/attachment/{fileid}`. Anyways, I am still not able to complete to do a PR and contribute. But hope this gives a headstart and helps in someway.

**Comment by yfrindel** [comment]
Personaly, on my side, I download the tests steps like that : 

```
public async void DownloadXml(string key,string _XmlFolderPath)
{
            var listTagsToClose = new List<string>();
            listTagsToClose.Add("statusCategory");
            listTagsToClose.Add("attachment");

            var listTagsToEscape = new List<string>();
            listTagsToEscape.Add("title");
            listTagsToEscape.Add("summary");
            listTagsToEscape.Add("iconUrl");
            listTagsToEscape.Add("name");

            String URLString = String.Format("http://jiralink.com/si/jira.issueviews:issue-xml/{0}/{0}.xml?os_username={1}&os_password={2}", key, this.Settings.UserID, this.Settings.Pass);
            XmlTextReader reader = new XmlTextReader(URLString);
            
            XmlUrlResolver resolver = new XmlUrlResolver();
            reader.XmlResolver = resolver;
           
            // Parse the file.
            using (System.IO.StreamWriter file =
                new System.IO.StreamWriter(_XmlFolderPath + key + ".xml"))
            {
                file.WriteLine("<?xml version=\"1.0\" encoding=\"utf-8\"?>");
                string lastReaderName = reader.Name;
                while (reader.Read())
                {
                    // Do any additional processing here.
                    switch (reader.NodeType)
                    {
                        case XmlNodeType.Comment:
                            file.WriteLine("<!-- " + reader.Value + " -->");
                            break;
                        case XmlNodeType.Element: // Le nœud est un élément.
                            file.Write("<" + reader.Name);

                            bool needToClose = false;

                            if (listTagsToClose.Contains(reader.Name))
                                 needToClose = true;

                            while (reader.MoveToNextAttribute()) // Lit les attributs.
                            {
                                string val2 = reader.Value;
                                if (listTagsToEscape.Contains(reader.Name))
                                    val2 = EscapeWrongCharacters(val2);
                                file.Write(" " + reader.Name + "='" + val2 + "'");
                            }

                            if (needToClose)
                                { file.Write("/>"); }
                            else
                                file.Write(">");
                            //file.WriteLine("");
                            break;
                        case XmlNodeType.Text: //Affiche le texte de chaque élément.
                            string val = reader.Value;
                            if (listTagsToEscape.Contains(lastReaderName))
                                val = EscapeWrongCharacters(val);
                            file.Write(val);
                            break;
                        case XmlNodeType.EndElement: //Affiche la fin de l'élément.
                            file.Write("</" + reader.Name);
                            file.WriteLine(">");
                            break;
                        case XmlNodeType.CDATA:
                            file.WriteLine("![CDATA[ " + EscapeWrongCharacters(reader.Value) + " ]]");
                            break;
                    }
                    lastReaderName = reader.Name;
                }               
            }
            // Close the reader.
            reader.Close();

            // Try to open XML
            try
            {
                XDocument document = XDocument.Load(_XmlFolderPath + key + ".xml");
            }
            catch (Exception ex)
            {
                Logger.Log(LogLevel.Warning, $"The XML file "+_XmlFolderPath + key + ".xml seems not valid for " + key + " : "+ex.InnerException);
            }
            
}
```

then the xml downloaded with the tests descriptions are imported like that in Azure devops : 

```
NetworkCredential netCred = new NetworkCredential("yannick.blabla@company.com", token);
BasicAuthCredential basicCred = new BasicAuthCredential(netCred);
TfsClientCredentials tfsCred = new TfsClientCredentials(basicCred);
tfsCred.AllowInteractive = false;
TfsTeamProjectCollection tfs = new TfsTeamProjectCollection(new Uri(url), tfsCred);           
tfs.Authenticate();
ITestManagementService service = (ITestManagementService)tfs.GetService(typeof(ITestManagementService));
ITestManagementTeamProject testProject = service.GetTeamProject(config.TargetProject);

//Test case update
foreach (WIToUpdate testUpdate in lstTests)
{
                Thread.Sleep(TimeSpan.FromSeconds(5));
                string filenamepath = xmlPath + testUpdate.originId + ".xml";
                if ((testUpdate.originId != "") && (File.Exists(filenamepath)))
                {
                    XmlDocument document = new XmlDocument();
                    document.Load(filenamepath);

                    // Description
                    XmlNode _nodeDescription = document.DocumentElement.SelectSingleNode("//channel/item/description");
                    String _nodeDescriptionTxt = "";
                    if (_nodeDescription != null)
                        _nodeDescriptionTxt = _nodeDescription.InnerXml;
                    
                    // Steps
                    XmlNodeList _nodeList = document.DocumentElement.SelectNodes("//customfieldvalues/steps/step");
                    List<TestStep> steps = new List<TestStep>();
                    foreach (XmlElement node in _nodeList)
                    {
                        string _step = RemoveWrongCharInStepText(node.SelectSingleNode("step").InnerText);
                        string _data = RemoveWrongCharInStepText(node.SelectSingleNode("data").InnerText);
                        string vstsAction = _step;
                        if (!string.IsNullOrEmpty(_data))
                            vstsAction += "\r\n" + _data;
                        string _result = RemoveWrongCharInStepText(node.SelectSingleNode("result").InnerText);

                        steps.Add(new TestStep(vstsAction, _result));
                    }

                    // Test Type
                    XmlNodeList _nodeTestTypeList = document.DocumentElement.SelectNodes("//customfields/customfield/customfieldname[contains(text(),'Test Type')]/../customfieldvalues/customfieldvalue");
                    string lstTestType = "";
                    foreach (XmlElement node in _nodeTestTypeList)
                    {
                        if (!string.IsNullOrEmpty(lstTestType))
                            lstTestType += ", ";
                        lstTestType += node.InnerText;
                    }
                    
                    ArrayList lstInvalid = null;

                    try
                    {
                        ITestCase testCase = testProject.TestCases.Find(testUpdate.wi.Id);
                        if (testCase.WorkItem.Tags.Contains("to_recycle"))
                            testCase.WorkItem.Tags = testCase.WorkItem.Tags.Replace("to_recycle", "");

                        testCase.Actions.Clear();
                        foreach (TestStep step in steps)
                        {
                            ITestStep newStep = testCase.CreateTestStep();
                            newStep.Title = step.title;
                            newStep.ExpectedResult = step.result;
                            testCase.Actions.Add(newStep);
                        }

                        if (!String.IsNullOrEmpty(lstTestType) && (lstTestType.Contains("Automated")))
                        {
                            if (!String.IsNullOrEmpty(testCase.WorkItem.Tags))
                                testCase.WorkItem.Tags += "; ";
                            testCase.WorkItem.Tags += "Automated";
                        }

                        lstInvalid = testCase.WorkItem.Validate();
                        testCase.Save();

                        // Update the description (for the text format)
                        Thread.Sleep(TimeSpan.FromSeconds(1));
                        WorkItem workItem = _store.GetWorkItem(testUpdate.wi.Id);
                        workItem.Open();
                        workItem.Fields["System.Description"].Value = _nodeDescriptionTxt;

                        if (!String.IsNullOrEmpty(lstTestType) && (lstTestType.Contains("Automated")))
                        {
                            workItem.Fields["Automation status"].Value = "Planned";
                        }

                            lstInvalid = workItem.Validate();
                        workItem.Save();

                        Logger.Log(LogLevel.Info, $"Steps for the test " + testUpdate.originId + "(" + testUpdate.wi.Id + ") updated.");
                    }
                    catch (Exception ex)
                    {
                        string lstFieldsError = "";
                        foreach (var item in lstInvalid)
                        {
                            if (!string.IsNullOrEmpty(lstFieldsError))
                                lstFieldsError += ", ";
                            lstFieldsError += ((Field)item).Name;
                        }
                        Logger.Log(LogLevel.Error, String.Format("Update failed for the Steps for the test {0} ({1}) : {2} (invalid fields : {3})", testUpdate.originId, testUpdate.wi.Id, ex.Message, lstFieldsError));                     
                    }                   
                }
}
```
